BibBuddy/app/src/main/java/de/bibbuddy/BookFragment.java

```diff
@@ -108,6 +125,199 @@ public boolean onOptionsItemSelected(MenuItem item) {
+   */
+  @Override
+  public void onRequestPermissionsResult(int requestCode,
+                                         String[] permissions, int[] grantResults) {
+
+    switch (requestCode) {
+
+      case STORAGE_PERMISSION_CODE:
+        if (grantResults.length > 0
+            && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+          createBibFile("Download", shelfName);
+          retrieveBibContent();
+          writeBibFile("Download", shelfName, bibExportContent);
+        } else {
+          Toast.makeText(getContext(), R.string.storage_permission_denied,
+              Toast.LENGTH_SHORT).show();
+        }
+        break;
+
+      default:
+    }
+  }
+
+  private void createBibFile(String folderName, String fileName) {
+    try {
+      String rootPath = Environment.getExternalStorageDirectory() + "/" + folderName + "/";
+      File root = new File(rootPath);
+
+      if (!root.exists()) {
+        root.mkdirs();
+      }
+
+      File file = new File(rootPath + fileName + ".bib");
+      if (file.exists()) {
+        file.delete();
+      }
+      file.createNewFile();
+
+      FileOutputStream out = new FileOutputStream(file);
+      out.flush();
+      out.close();
+
+      Toast.makeText(getContext(), getString(R.string.exported_file_stored_in) + '\n'
+          + "/" + folderName + "/" + fileName + ".bib", Toast.LENGTH_LONG).show();
+
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  private void retrieveBibContent() {
+
+    DatabaseHelper dbHelper = new DatabaseHelper(getContext());
+    BookDao bd = new BookDao(dbHelper);
+    NoteDao nd = new NoteDao(dbHelper);
+
+    List<Long> bookIdsCurrShelf = bd.getAllBookIdsForShelf(shelfId);
+    List<Author> authorsCurrBook = new ArrayList<>();
+    List<Long> notesCurrBook = new ArrayList<>();
+    String allNotesCurrBook = "";
+    String authorNamesCurrBook = "";
+
+    //for each book in the current shelf
+    for (int i = 0; i < bookIdsCurrShelf.size(); i++) {
+      Book currBook;
+      Long currBookId = bookIdsCurrShelf.get(i);
+      authorsCurrBook = bd.getAllAuthorsForBook(currBookId);
+      notesCurrBook = nd.getAllNoteIdsForBook(currBookId);
+      currBook = bd.findById(currBookId);
+
+      /*
+      get the notes for the current book
+      and save the bib content in one string
+      */
+      for (int k = 0; k < notesCurrBook.size(); k++) {
+        String noteTextCurrBook = nd.findTextById(notesCurrBook.get(k));
+        allNotesCurrBook +=  "anote={" + noteTextCurrBook + "}," + '\n';
+      }
+
+      /*
+      get author's first and last name and include
+      the needed book data in a bib format
+      */
+      for (int u = 0; u < authorsCurrBook.size(); u++) {
+        authorNamesCurrBook = authorsCurrBook.get(u).getFirstName()
+            + " " + authorsCurrBook.get(u).getLastName();
+      }
+
+      bibExportContent = bibExportContent
+          + "@book{" + currBook.getTitle() + currBook.getPubYear() + "," + '\n'
```

Comment:
The problem here is that the title can have blank spaces, but the bibtex-id of the item is not allowed to have. I would therefore suggest to either strip spaces or, which I think is better because it is shorter, use only the authors lastname and the year.

Later on, we should also check, if this ID was already used, as this may throw errors when further used.

```suggestion
          + "@book{" + authorsCurrBook[0].getLastName().replaceAll("\\s+","") + currBook.getPubYear() + "," + '\n'
```

Reply:
If you for example have 2 books from the same author, written in the same year, then the keys for these books will be the same. This should be a unique key for the book.

Reply:
True, that's why I wrote that we need a check for that later on - you can of course implement that now. Or use the title, but then please trim the whitespace. However, the check will be needed nevertheless (book with same title and year can also occur).

Reply:
ok